src/t_zset.c

```diff
@@ -4019,9 +4041,14 @@ void genericZpopCommand(client *c, robj **keyv, int keyc, int where, int emitkey
                 ele = sdsnewlen(vstr,vlen);
 
             /* Get the score. */
-            sptr = lpNext(zl,eptr);
             serverAssertWithInfo(c,zobj,sptr != NULL);
             score = zzlGetScore(sptr);
+            /* In previous versions, we call notifyKeyspaceEvent after the first entry is deleted. 
```

Comment:
I think we can move the `notifyKeyspaceEvent` to after all the pop completed. This way, we won't need to check the `result_count`, and the logic will be much clearer. Commands like `zremrange*` operate in the same manner since the execution of the command is atomic.

module is a bit special, but I believe that moving the notification from after the first element is popped to after all elements have been popped will not break compatibility with modules.

I'd also like to ping @oranagra to confirm this.

Reply:
indeed the logic of firing the notification only on the first removal seems odd (rather than firing on every removal).
when we realize that when this code was created in 56bbab238a there were no modules, then firing it on the first removal was equivalent to firing it at the end (if there was at least one removal).
but with modules it makes less sense.
i'd vote to move this to the end and document the behavior change for modules.
@MeirShpilraien WDYT?